semantics/src/main/kotlin/com/strumenta/kolasu/semantics/indexing/IndexWriter.kt

```diff
@@ -0,0 +1,66 @@
+package com.strumenta.kolasu.semantics.indexing
+
+import com.strumenta.kolasu.utils.sortBySubclassesFirst
+import kotlin.reflect.KClass
+import kotlin.reflect.safeCast
+
+/**
+ * Type-safe builder to create index writer instances.
+ * @param configuration the index writer configuration
+ * @return the configured index writer instance
+ **/
+fun indexWriter(
+    index: Index = InMemoryIndex(),
+    configuration: IndexWriterConfigurator.() -> Unit
+) = mutableMapOf<KClass<*>, (Any) -> Any?>()
+    .also { rules -> IndexWriterConfigurator(rules).apply(configuration) }
+    .let { rules -> IndexWriter(index, rules) }
+
+/**
+ * Configurable component handling write operations over an index.
+ * @property index the underlying index
+ * @property rules the index writer rules to use
+ **/
+class IndexWriter(
+    private val index: Index = InMemoryIndex(),
+    private val rules: Map<KClass<*>, (Any) -> Any?> = emptyMap()
+) {
+    /**
+     * Execute the index writer rule for the given [input].
+     * @param input the index writer rule input
+     * @return the output of the corresponding rule
```

Comment:
Does null indicate that no symbol has been produced for the given node?

Reply:
null is returns in two cases:

* there is no rule for the given input
* the rule produced a null value

A node is written into the index iff a rules exists and it produces a non-null value. The second case provides the possibility to have conditional rules - e.g. export classes only if public.

I will update the method documentation to make this clear.

Reply:
> I will update the method documentation to make this clear.

Good, thank you

Regarding IndexWriter, I have the same comments I have for IndexReader

Reply:
After discussion we decided to call this `RepositoryWriter`